### PR TITLE
make-initrd-ng: use goblin instead of shelling out to patchelf and friends

### PR DESCRIPTION
**Pull request by K900**
###### Description of changes

Should be faster and also fix cross maybe?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/systemd/initrd.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd/initrd.nix:172]: This doesn't appear to be getting passed to the `makeInitrdNG`, does it?

**Review by ElvishJerricco — COMMENTED** [review]
You can probably squash out the llvm-strip commit, right?

File: nixos/modules/system/boot/systemd/initrd.nix
Comment (K900) [reply nixos/modules/system/boot/systemd/initrd.nix:172]: Looks like it. ping @lheckemann

**Comment by K900** [comment]
Rebased out the llvm-strip commit.

File: nixos/modules/system/boot/systemd/initrd.nix
Comment (lheckemann) [reply nixos/modules/system/boot/systemd/initrd.nix:172]: whooops yes, accidentally left that unstaged.

Comment (lheckemann) [reply nixos/modules/system/boot/systemd/initrd.nix:172]: https://github.com/lheckemann/nixpkgs/commit/c9b29fccbf5b132a7b0f7b61a7404622482d3b64

Comment (K900) [reply nixos/modules/system/boot/systemd/initrd.nix:173]: Picked

**Review by ElvishJerricco — APPROVED** [review]
On the condition that we are OK with adding external rust deps to `make-initrd-ng`, LGTM. I'll let others decide on that, because I'm finding it too difficult to decide.

**Comment by K900** [comment]
Just to reiterate my arguments in favor of this on Matrix:

- `goblin` isn't some random crate, it's actively maintained and used by a whole lot of people, including Sentry, Maturin, a bunch of crypto projects and many others
- The deps are pure Rust, so we're not really increasing closure size or bootstrap complexity by a relevant amount
- Keeping the deps up to date will be much easier when make-initrd-ng grows out of nixpkgs and into its own repo, which I'm expecting to happen sooner rather than later

**Comment by lheckemann** [comment]
>Keeping the deps up to date

https://github.com/lheckemann/nixpkgs/commit/cc635373df45e1959213bc7b1beb81b901387f29 ? (cc @dasJ @ElvishJerricco I took the liberty of adding you as maintainers in this commit, let me know if you don't want that)

File: nixos/modules/system/boot/systemd/initrd.nix
Comment (SuperSandro2000) [reply nixos/modules/system/boot/systemd/initrd.nix:173]: Why do we need an option for this at all? When would I want to turn this off?

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (SuperSandro2000) [inline pkgs/build-support/kernel/make-initrd-ng.nix:80]: This should be using lib.optionalString instead of attr merging.

**Comment by SuperSandro2000** [comment]
> On the condition that we are OK with adding external rust deps to `make-initrd-ng`

That's fine

File: nixos/modules/system/boot/systemd/initrd.nix
Comment (K900) [reply nixos/modules/system/boot/systemd/initrd.nix:173]: As the description says, it's a pretty small initrd size increase for a much bigger build closure decrease, so some people might want to take that trade.

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (ElvishJerricco) [reply pkgs/build-support/kernel/make-initrd-ng.nix:80]: No it shouldn't. It needs to be *unset* for the program to work, not empty.

Comment (K900) [reply pkgs/build-support/kernel/make-initrd-ng.nix:80]: I think setting it to `null` would make it unset in the stdenv, but don't quote me on that.

**Comment by ElvishJerricco** [comment]
@SuperSandro2000 Can you elaborate on why you think it's fine? Just for completeness, I'll reiterate my concerns with adding external dependencies.

- We have to keep dependencies up to date, which is historically not done well unless automated.
- It adds more external code that we have to trust is implemented correctly.
- Readers must now seek external documentation to contribute to or understand large parts of the `make-initrd-ng` program. It is no longer self-contained.
- Building `make-initrd-ng` now takes ~12s, compared to < 1s before. This greatly annoys me while iterating on changes to the program. (Maybe we can gets deps built in a separate derivation?)

This isn't to say that these concerns are more significant than (or even as accurate as) the benefits described by @K900. There are even some benefits they haven't mentioned, such as the fact that `make-initrd-ng` now takes literally a tenth as long to run, even less if you disable stripping (though it was already ~1s before). I'm saying that I haven't seen it convincingly argued one way or the other.

I would like to see a clear and well-thought-out argument, even if only a brief one, from someone other than @K900 or myself that attempts to reconcile these concerns with the benefits before we merge.

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (ElvishJerricco) [reply pkgs/build-support/kernel/make-initrd-ng.nix:80]: @K900 That appears to be correct.

**Review by winterqt — CHANGES_REQUESTED** [review]
This is a nit, but we should also probably use a consistent commit naming scheme -- `make-initrd-ng` versus `makeInitrdNG`.

**Comment by ElvishJerricco** [comment]
> This is a nit, but we should also probably use a consistent commit naming scheme -- make-initrd-ng versus makeInitrdNG.

We want to avoid confusion between the Nix function `makeInitrdNG` (which creates a derivation that calls the tool, and is named after the old `makeInitrd`), and the tool itself. So its all-packages.nix attribute got the name `makeInitrdNGTool`.

**Comment by winterqt** [comment]
> We want to avoid confusion between the Nix function `makeInitrdNG` (which creates a derivation that calls the tool, and is named after the old `makeInitrd`), and the tool itself. So its all-packages.nix attribute got the name `makeInitrdNGTool`.

That's what I thought, but the strip commit *does* modify both -- but I do understand why it uses that name in its commit message now -- the optional stripping isn't an exclusive internal detail of the tool like Goblin versus `patchelf` is. At least, that's how it makes sense to me.

**Comment by K900** [comment]
The correct answer is "I am physically unable to keep track of all the ways we spell it".

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (K900) [reply pkgs/build-support/kernel/make-initrd-ng.nix]: Fixed.

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (K900) [reply pkgs/build-support/kernel/make-initrd-ng.nix]: Applied.

File: pkgs/build-support/kernel/make-initrd-ng.nix
Comment (ElvishJerricco) [inline pkgs/build-support/kernel/make-initrd-ng.nix:79]: As I said, this isn't right. It can't be an empty string. It has to be `null`.

Comment (K900) [reply pkgs/build-support/kernel/make-initrd-ng.nix]: Ugh. Forgot that's not what `optionalString` does.

**Comment by SuperSandro2000** [comment]
> * We have to keep dependencies up to date, which is historically not done well unless automated.

Yeah, we should probably configure dependabot for those libraries or move them into their own repository if they get big enough.

> * It adds more external code that we have to trust is implemented correctly.

I don't think that is different to adding new libraries to projects we don't develop. There is not really a standard about that either and usually we just do it.

> * Readers must now seek external documentation to contribute to or understand large parts of the `make-initrd-ng` program. It is no longer self-contained.

That is very similar for a lot of libraries we are already using. 

> * Building `make-initrd-ng` now takes ~12s, compared to < 1s before. This greatly annoys me while iterating on changes to the program. (Maybe we can gets deps built in a separate derivation?)

12 seconds is still quite fast and we didn't boost the compile time to minutes.
If the standard rust tooling allows to do that, sure why not but 12 seconds is really not long.

---

We should try to keep dependencies minimal and only use them where it makes sense and the library saves us a lot of code duplication. Especially when dealing with complex topics we can save us a lot of development time by using majored libraries. Adding a crate for isEven/isOdd would be overkill.

**Comment by winterqt** [comment]
@ofborg test systemd-initrd-simple

File: pkgs/build-support/kernel/make-initrd-ng-tool.nix
Comment (SuperSandro2000) [inline pkgs/build-support/kernel/make-initrd-ng-tool.nix:10]: I don't think that will work if version is not updated.

Comment (lheckemann) [reply pkgs/build-support/kernel/make-initrd-ng-tool.nix:10]: It did for meshcentral: https://github.com/NixOS/nixpkgs/pull/164212

Comment (SuperSandro2000) [reply pkgs/build-support/kernel/make-initrd-ng-tool.nix:10]: Oh, really. Didn't know that.

**Review by ElvishJerricco — APPROVED** [review]

**Pull request closed without merging**
